R/simulate.r

```diff
@@ -1,3 +1,309 @@
+#' International journal of environmental research and public health, 7(3),
+#' 1186–1204. \doi{https://doi.org/10.3390/ijerph7031204}
+#'
+#' Blumberg, S., and J. O. Lloyd-Smith. 2013. "Comparing Methods for
+#' Estimating R0 from the Size Distribution of Subcritical Transmission
+#' Chains." Epidemics 5 (3): 131–45.
+#' \doi{https://doi.org/10.1016/j.epidem.2013.05.002}.
+#'
+#' Farrington, C. P., M. N. Kanaan, and N. J. Gay. 2003.
+#' "Branching Process Models for Surveillance of Infectious Diseases
+#' Controlled by Mass Vaccination.” Biostatistics (Oxford, England)
+#' 4 (2): 279–95. \doi{https://doi.org/10.1093/biostatistics/4.2.279}.
+simulate_chains <- function(
+    index_cases,
+    statistic = c("size", "length"),
+    offspring_dist,
+    ...,
+    stat_max = Inf,
+    pop = Inf,
+    percent_immune = 0,
+    generation_time = NULL,
+    t0 = 0,
+    tf = Inf) {
+  # Input checking
+  checkmate::assert_count(index_cases, positive = TRUE)
+  checkmate::assert_choice(statistic, choices = c("size", "length"))
+  checkmate::assert_string(offspring_dist)
+  # check that offspring function exists in the environment
+  roffspring_name <- paste0("r", offspring_dist)
+  check_offspring_func_valid(roffspring_name)
+  checkmate::assert(
+    is.infinite(stat_max) ||
+      checkmate::assert_integerish(stat_max, lower = 0)
+  )
+  checkmate::assert(
+    is.infinite(pop) ||
+      checkmate::assert_integerish(pop, lower = 0)
+  )
+  checkmate::assert_number(
+    percent_immune,
+    lower = 0, upper = 1
+  )
+  if (!missing(generation_time)) {
+    check_generation_time_valid(generation_time)
+  } else if (!missing(tf)) {
+    stop("If `tf` is specified, `generation_time` must be specified too.")
+  }
+  checkmate::assert_numeric(
+    t0,
+    lower = 0, finite = TRUE
+  )
+  checkmate::assert_number(
+    tf,
+    lower = 0
+  )
+  # Gather offspring distribution parameters
+  pars <- list(...)
+
+  # Initialisations
+  stat_track <- rep(1, index_cases) # track statistic (length or size)
+  n_offspring <- rep(1, index_cases) # current number of offspring
+  sim <- seq_len(index_cases) # track trees that are still being simulated
+  infector_ids <- rep(1, index_cases)
+
+  # initialise list of data frame to hold the transmission trees
+  generation <- 1L
+  tree_df <- list(
+    data.frame(
+      infectee_id = seq_len(index_cases),
+      sim_id = 1L,
+      infector_id = NA_integer_,
+      generation = generation
+    )
+  )
+  # Initialise susceptible population
+  # I think this currently implicitly assumes the initial cases are imported
+  # (not subtracted from pop). If not, we'll have to subtract the non-imported
+  # cases. An improvement of this could take the form of an extra argument that
+  # defines a percentage of initial cases suspected to be imported. We can
+  # then substract only the non-imported cases from pop.
+  susc_pop <- round(pop * (1 - percent_immune))
```

Comment:
My initial thought is that it would be fine to just subtract the index cases (and make sure it doesn't go <0) but I might not have all use cases in mind.

```suggestion
  susc_pop <- max(round(pop * (1 - percent_immune)) - index_cases, 0)
```

Reply:
Agreed. Let's go with this suggestion for now.